### ISSUE TITLE
Migrating from v2 -> v3 config (currently running envoy v1.16.0, want to upgrade to latest)

### ISSUE THREAD

**Issue by sharkannon** [envoyproxy/envoy#24276]
Currently running v1.16.0 of envoy and looking to upgrade it, but having a hard time finding any documentation that really explains the differences and upgrade path from v2 -> v3.  Maybe I'm blind, but that's me.  Our current config is:

```
#/etc/envoy/envoy.yaml
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 8095
static_resources:
  listeners:
    - name: listener_0
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 8094
      filter_chains:
        - filters:
            - name: envoy.http_connection_manager
              config:
                stat_prefix: ingress_http
                use_remote_address: true
                stream_idle_timeout: 0s
                access_log:
                  - name: envoy.file_access_log
                    config:
                      path: /dev/stdout
                      format: >
                        [%START_TIME%] "%REQ(:METHOD)%
                        %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"
                        %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED%
                        %BYTES_SENT% %DURATION%
                        %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%
                        "%REQ(X-FORWARDED-FOR)%" "%REQ(USER-AGENT)%"
                        "%REQ(X-TRACE)%" "%REQ(X-CUSTOMER-ID)%"
                        "%REQ(:AUTHORITY)%" "%UPSTREAM_HOST%"
                  - name: envoy.http_grpc_access_log
                    config:
                      common_config:
                        log_name: apirate
                        grpc_service:
                          envoy_grpc:
                            cluster_name: ratelimit
                      additional_request_headers_to_log:
                        - x-customer-id
                        - x-trace
                        - x-forwarded-for
                        - user-agent
                route_config:
                  name: local_route
                  request_headers_to_add:
                    - header:
                        key: x-forwarded-proto
                        value: https
                      append: false
                    - header:
                        key: connection
                        value: close
                      append: false
                  virtual_hosts:
                    - name: local_service
                      domains:
                        - '*'
                      routes:
                        - match:
                            prefix: /rest/transport/
                            headers:
                              - name: X-HAPROXY-CLUSTER
                                exact_match: kube
                          route:
                            cluster: backend-kube
                            timeout: 0s
                            retry_policy:
                              retry_on: connect-failure
                              num_retries: 5
                              retry_host_predicate:
                                - name: envoy.retry_host_predicates.previous_hosts
                              host_selection_retry_max_attempts: 5
                        - match:
                            prefix: /
                            headers:
                              - name: X-HAPROXY-CLUSTER
                                exact_match: kube
                          route:
                            cluster: backend-kube
                            timeout: 0s
                            retry_policy:
                              retry_on: connect-failure
                              num_retries: 5
                              retry_host_predicate:
                                - name: envoy.retry_host_predicates.previous_hosts
                              host_selection_retry_max_attempts: 5
                            rate_limits:
                              actions:
                                - generic_key:
                                    descriptor_value: apirate
                                - request_headers:
                                    header_name: ':authority'
                                    descriptor_key: authority
                                - request_headers:
                                    header_name: ':method'
                                    descriptor_key: method
                                - request_headers:
                                    header_name: ':path'
                                    descriptor_key: path
                                - request_headers:
                                    header_name: user-agent
                                    descriptor_key: userAgent
                                - request_headers:
                                    header_name: x-customer-id
                                    descriptor_key: customerId
                                - request_headers:
                                    header_name: x-trace
                                    descriptor_key: xTrace
                                - request_headers:
                                    header_name: x-forwarded-for
                                    descriptor_key: xForwardedFor
                http_filters:
                  - name: envoy.rate_limit
                    config:
                      domain: apirate
                      failure_mode_deny: false
                      timeout: 0.25s
                      rate_limit_service:
                        grpc_service:
                          timeout: 0.05s
                          envoy_grpc:
                            cluster_name: ratelimit
                  - name: envoy.router
  clusters:
    - name: backend-kube
      connect_timeout: 0.25s
      type: STRICT_DNS
      lb_policy: ROUND_ROBIN
      max_requests_per_connection: 1
      http_protocol_options: {}
      dns_refresh_rate: 5s
      load_assignment:
        cluster_name: backend-kube
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 81
      health_checks:
        http_health_check:
          path: /healthz
        timeout: 1s
        interval: 1s
        interval_jitter_percent: 25
        unhealthy_threshold: 3
        healthy_threshold: 1
        reuse_connection: true
        always_log_health_check_failures: true
      circuit_breakers:
        thresholds:
          - priority: DEFAULT
            max_connections: 65536
            max_pending_requests: 65536
            max_requests: 65536
            max_retries: 1024
          - priority: HIGH
            max_connections: 65536
            max_pending_requests: 65536
            max_requests: 65536
            max_retries: 1024
    - name: ratelimit
      connect_timeout: 0.25s
      type: STRICT_DNS
      lb_policy: ROUND_ROBIN
      http2_protocol_options: {}
      load_assignment:
        cluster_name: ratelimit
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 192.168.0.1
                      port_value: 8081
```

I'd like to upgrade that to v3 format, but get errors like `INVALID_ARGUMENT:(static_resources.listeners[0].filter_chains[0].filters[0]) config: Cannot find field.) has unknown fields`, which though I'm aware of config being depricated, I know type_config is available.. but not 100% sure how all this works.  Any help (with examples) would be appreciated.

**Comment by kyessenov** [comment]
Yeah, the biggest thing is the usage of the "typed_config" consistently:

```
 - name: envoy.http_connection_manager
   typed_config:
     "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
     stat_prefix: ingress_http
```

There were some field renames that can be found by looking at the protobuf annotations, so you probably want to gradually convert config and enable strict validation (disallow unknown fields in static and dynamic configs).

**Comment by sharkannon** [comment]
I'm not really impressed with the logging messages about the deprications, but I *think* I got it.  I basically had to fix deprication warnings in 1.16, then upgrade toi 1.17, then fix those.. all the way up to 1.21.. then in 1.22 there was a couple of those type_config issues that came up...


```
admin:
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 8095
static_resources:
  listeners:
    - name: listener_0
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 8094
      filter_chains:
        - filters:
            - name: envoy.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                stat_prefix: ingress_http
                use_remote_address: true
                stream_idle_timeout: 0s
                access_log:
                  - name: envoy.access_loggers
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.access_loggers.stream.v3.StdoutAccessLog
                      log_format:
                        text_format_source: 
                          inline_string: >
                            [%START_TIME%] "%REQ(:METHOD)%
                            %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"
                            %RESPONSE_CODE% %RESPONSE_FLAGS% %BYTES_RECEIVED%
                            %BYTES_SENT% %DURATION%
                            %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%
                            "%REQ(X-FORWARDED-FOR)%" "%REQ(USER-AGENT)%"
                            "%REQ(X-TRACE)%" "%REQ(X-CUSTOMER-ID)%"
                            "%REQ(:AUTHORITY)%" "%UPSTREAM_HOST%"
                  - name: envoy.access_loggers
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.access_loggers.grpc.v3.HttpGrpcAccessLogConfig
                      common_config:
                        log_name: apirate
                        transport_api_version: V3
                        grpc_service:
                          envoy_grpc:
                            cluster_name: ratelimit
                      additional_request_headers_to_log:
                        - x-customer-id
                        - x-trace
                        - x-forwarded-for
                        - user-agent
                route_config:
                  name: local_route
                  request_headers_to_add:
                    - header:
                        key: x-forwarded-proto
                        value: https
                      append: false
                    - header:
                        key: connection
                        value: close
                      append: false
                  virtual_hosts:
                    - name: local_service
                      domains:
                        - '*'
                      routes:
                        - match:
                            prefix: /rest/transport/
                            headers:
                              - name: X-HAPROXY-CLUSTER
                                string_match: 
                                  exact: kube
                          route:
                            cluster: backend-kube
                            timeout: 0s
                            retry_policy:
                              retry_on: connect-failure
                              num_retries: 5
                              retry_host_predicate:
                                - name: envoy.retry_host_predicates
                                  typed_config:
                                    "@type": type.googleapis.com/envoy.extensions.retry.host.previous_hosts.v3.PreviousHostsPredicate
                              host_selection_retry_max_attempts: 5
                        - match:
                            prefix: /
                            headers:
                              - name: X-HAPROXY-CLUSTER
                                string_match: 
                                  exact: kube
                          route:
                            cluster: backend-kube
                            timeout: 0s
                            retry_policy:
                              retry_on: connect-failure
                              num_retries: 5
                              retry_host_predicate:
                                - name: envoy.retry_host_predicates
                                  typed_config:
                                    "@type": type.googleapis.com/envoy.extensions.retry.host.previous_hosts.v3.PreviousHostsPredicate
                              host_selection_retry_max_attempts: 5
                            rate_limits:
                              actions:
                                - generic_key:
                                    descriptor_value: apirate
                                - request_headers:
                                    header_name: ':authority'
                                    descriptor_key: authority
                                - request_headers:
                                    header_name: ':method'
                                    descriptor_key: method
                                - request_headers:
                                    header_name: ':path'
                                    descriptor_key: path
                                - request_headers:
                                    header_name: user-agent
                                    descriptor_key: userAgent
                                - request_headers:
                                    header_name: x-customer-id
                                    descriptor_key: customerId
                                - request_headers:
                                    header_name: x-trace
                                    descriptor_key: xTrace
                                - request_headers:
                                    header_name: x-forwarded-for
                                    descriptor_key: xForwardedFor
                http_filters:
                  - name: envoy.filters.http.ratelimit
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.ratelimit.v3.RateLimit
                      domain: apirate
                      failure_mode_deny: false
                      timeout: 0.25s
                      rate_limit_service:
                        transport_api_version: V3
                        grpc_service:
                          timeout: 0.05s
                          envoy_grpc:
                            cluster_name: ratelimit
                  - name: envoy.filters.http.router
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
  clusters:
    - name: backend-kube
      connect_timeout: 0.25s
      type: STRICT_DNS
      lb_policy: ROUND_ROBIN
      typed_extension_protocol_options:
        envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
          "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
          common_http_protocol_options:
            max_requests_per_connection: 1
          use_downstream_protocol_config: {}
      dns_refresh_rate: 5s
      load_assignment:
        cluster_name: backend-kube
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 127.0.0.1
                      port_value: 81
      health_checks:
        http_health_check:
          path: /healthz
        timeout: 1s
        interval: 1s
        interval_jitter_percent: 25
        unhealthy_threshold: 3
        healthy_threshold: 1
        reuse_connection: true
        always_log_health_check_failures: true
      circuit_breakers:
        thresholds:
          - priority: DEFAULT
            max_connections: 65536
            max_pending_requests: 65536
            max_requests: 65536
            max_retries: 1024
          - priority: HIGH
            max_connections: 65536
            max_pending_requests: 65536
            max_requests: 65536
            max_retries: 1024

    - name: ratelimit
      connect_timeout: 0.25s
      type: STRICT_DNS
      lb_policy: ROUND_ROBIN
      load_assignment:
        cluster_name: ratelimit
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: 192.168.0.1
                      port_value: 8081
```

That starts up, is it the same as my original, I'm not 100% sure.. but it at least starts up, I'll deploy the changes to dev tomorrow to see if it works.  Any suggestions/recommendations/fixes would be appreciated.

**Comment by kyessenov** [comment]
I don't see anything wrong but it's always worth testing before deploying. One piece of good news for you is that v3 is forever, there'll be no v4 migration.
